### ISSUE TITLE
Container does not shut down cleanly when docker stop is issued

### ISSUE THREAD

**Issue by Ageraluon** [bitnami/bitnami-docker-postgresql#79]
**Description**

I'm starting the postgresql container with a mounted volume. After the first initialization, I'm closing the container with 'docker stop', 'docker rm', then recreating the container with the same mounted volume (in my case a cifs mount from a windows host). The following output shows that the postgresql container has to go into recovery mode since the docker stop shutdown was not performed cleanly:

```
 Welcome to the Bitnami postgresql container
 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-postgresql
 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-postgresql/issues

 nami    INFO  Initializing postgresql
 postgre INFO  ==> No injected configuration files found. Creating default config file...
 postgre INFO  ==> Deploying PostgreSQL with persisted data...
 postgre INFO  ==> Configuring PostgreSQL...
 postgre INFO  ==> Configuring replication parameters...
 postgre INFO  ==> Configuring permissions for config files...
 postgre INFO
 postgre INFO  ########################################################################
 postgre INFO   Installation parameters for postgresql:
 postgre INFO     Persisted data and properties have been restored.
 postgre INFO     Any input specified will not take effect.
 postgre INFO   This installation requires no credentials.
 postgre INFO  ########################################################################
 postgre INFO
 nami    INFO  postgresql successfully initialized
 INFO  ==> Starting postgresql...
 2018-06-18 08:06:10.753 GMT [54] LOG:  listening on IPv4 address "0.0.0.0", port 5432
 2018-06-18 08:06:10.753 GMT [54] LOG:  listening on IPv6 address "::", port 5432
 2018-06-18 08:06:10.819 GMT [54] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
 2018-06-18 08:06:10.952 GMT [56] LOG:  database system was interrupted while in recovery at 2018-06-18 08

 2018-06-18 08:06:10.952 GMT [56] HINT:  This probably means that some data is corrupted and you will have
st backup for recovery.
 2018-06-18 08:06:13.872 GMT [56] LOG:  could not fsync file "./base/1": Invalid argument
 2018-06-18 08:06:14.895 GMT [56] LOG:  could not fsync file "./base/12293": Invalid argument
 2018-06-18 08:06:15.913 GMT [56] LOG:  could not fsync file "./base/12294": Invalid argument
 2018-06-18 08:06:15.913 GMT [56] LOG:  could not fsync file "./base": Invalid argument
 2018-06-18 08:06:16.121 GMT [56] LOG:  could not fsync file "./global": Invalid argument
 2018-06-18 08:06:16.125 GMT [56] LOG:  could not fsync file "./pg_commit_ts": Invalid argument
 2018-06-18 08:06:16.128 GMT [56] LOG:  could not fsync file "./pg_dynshmem": Invalid argument
 2018-06-18 08:06:16.156 GMT [56] LOG:  could not fsync file "./pg_logical/mappings": Invalid argument
 2018-06-18 08:06:16.160 GMT [56] LOG:  could not fsync file "./pg_logical/snapshots": Invalid argument
 2018-06-18 08:06:16.160 GMT [56] LOG:  could not fsync file "./pg_logical": Invalid argument
 2018-06-18 08:06:16.170 GMT [56] LOG:  could not fsync file "./pg_multixact/members": Invalid argument
 2018-06-18 08:06:16.177 GMT [56] LOG:  could not fsync file "./pg_multixact/offsets": Invalid argument
 2018-06-18 08:06:16.177 GMT [56] LOG:  could not fsync file "./pg_multixact": Invalid argument
 2018-06-18 08:06:16.194 GMT [56] LOG:  could not fsync file "./pg_notify": Invalid argument
 2018-06-18 08:06:16.197 GMT [56] LOG:  could not fsync file "./pg_replslot": Invalid argument
 2018-06-18 08:06:16.202 GMT [56] LOG:  could not fsync file "./pg_serial": Invalid argument
 2018-06-18 08:06:16.211 GMT [56] LOG:  could not fsync file "./pg_snapshots": Invalid argument
 2018-06-18 08:06:16.216 GMT [56] LOG:  could not fsync file "./pg_stat": Invalid argument
 2018-06-18 08:06:16.236 GMT [56] LOG:  could not fsync file "./pg_stat_tmp": Invalid argument
 2018-06-18 08:06:16.250 GMT [56] LOG:  could not fsync file "./pg_subtrans": Invalid argument
 2018-06-18 08:06:16.257 GMT [56] LOG:  could not fsync file "./pg_tblspc": Invalid argument
 2018-06-18 08:06:16.267 GMT [56] LOG:  could not fsync file "./pg_twophase": Invalid argument
 2018-06-18 08:06:16.288 GMT [56] LOG:  could not fsync file "./pg_wal/archive_status": Invalid argument
 2018-06-18 08:06:16.288 GMT [56] LOG:  could not fsync file "./pg_wal": Invalid argument
 2018-06-18 08:06:16.297 GMT [56] LOG:  could not fsync file "./pg_xact": Invalid argument
 2018-06-18 08:06:16.317 GMT [56] LOG:  could not fsync file ".": Invalid argument
 2018-06-18 08:06:16.320 GMT [56] LOG:  could not fsync file "pg_tblspc": Invalid argument
 2018-06-18 08:06:16.361 GMT [56] LOG:  database system was not properly shut down; automatic recovery in

 2018-06-18 08:06:16.489 GMT [56] LOG:  redo starts at 0/15C0260
 2018-06-18 08:06:16.491 GMT [56] LOG:  invalid record length at 0/15C0298: wanted 24, got 0
 2018-06-18 08:06:16.491 GMT [56] LOG:  redo done at 0/15C0260
 2018-06-18 08:06:16.568 GMT [56] FATAL:  could not fsync file "base/1": Invalid argument
 2018-06-18 08:06:16.572 GMT [54] LOG:  startup process (PID 56) exited with exit code 1
 2018-06-18 08:06:16.572 GMT [54] LOG:  aborting startup due to startup process failure
 2018-06-18 08:06:16.597 GMT [54] LOG:  database system is shut down
```

**Steps to reproduce the issue:**

1. run the postgresql container with a volume mounted into /bitnami
2. after initialization send docker stop, docker rm
3. run the postgresql container again with the same volume mounted

**Describe the results you received:**

The container does not send any message about a SIGTERM signal received and graceful shutdown. Instead the container is shut down immediately. This leads to the error above

**Describe the results you expected:**

container should start up normally.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Version**

- Output of `docker version`:

```
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:12:48 2018
 OS/Arch:      windows/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:22:38 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

- Output of `docker info`:

```
Containers: 4
 Running: 3
 Paused: 0
 Stopped: 1
Images: 99
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.87-linuxkit-aufs
Operating System: Docker for Windows
OSType: linux
Architecture: x86_64
CPUs: 3
Total Memory: 12.19GiB
Name: linuxkit-00155da802b3
ID: FAD6:TSRG:Z4YN:KYUT:WUO5:WE7G:XPRR:N3SF:6PNN:MXCB:TWP6:WQ4A
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 43
 Goroutines: 57
 System Time: 2018-06-18T08:16:47.9626354Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.21.1, build 7641a569
docker-py version: 3.2.1
CPython version: 3.6.4
OpenSSL version: OpenSSL 1.0.2k  26 Jan 2017
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**

**Comment by juan131** [comment]
Hi @Ageraluon 

I was unable to reproduce the issue using the `docker-compose.yml` file:

```
$ docker-compose up -d
$ docker-compose stop
$ docker-compose up -d 
$ docker-compose logs -f postgresql
postgresql_1  | INFO  ==> Starting postgresql...
postgresql_1  | 2018-06-18 10:27:22.612 GMT [87] LOG:  database system is ready to accept connections
postgresql_1  | 2018-06-18 10:29:14.803 GMT [33] LOG:  listening on IPv4 address "0.0.0.0", port 5432
postgresql_1  | 2018-06-18 10:29:14.803 GMT [33] LOG:  listening on IPv6 address "::", port 5432
postgresql_1  | 2018-06-18 10:29:14.808 GMT [33] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
postgresql_1  | 2018-06-18 10:29:14.820 GMT [35] LOG:  database system was interrupted; last known up at 2018-06-18 10:27:22 GMT
postgresql_1  | 2018-06-18 10:29:14.955 GMT [35] LOG:  database system was not properly shut down; automatic recovery in progress
postgresql_1  | 2018-06-18 10:29:14.958 GMT [35] LOG:  redo starts at 0/15C06B0
postgresql_1  | 2018-06-18 10:29:14.958 GMT [35] LOG:  invalid record length at 0/15C06E8: wanted 24, got 0
postgresql_1  | 2018-06-18 10:29:14.958 GMT [35] LOG:  redo done at 0/15C06B0
postgresql_1  | 2018-06-18 10:29:14.967 GMT [33] LOG:  database system is ready to accept connections
```

What directory are you mounting your volume at? 
Were you doing any kind of operation on the database when you stopped the container?

**Comment by Ageraluon** [comment]
Hi @juan131,

Thanks for trying it out. It seems you are experiencing the very problem I am currently facing:

```
postgresql_1  | 2018-06-18 10:29:14.820 GMT [35] LOG:  database system was interrupted; last known up at 2018-06-18 10:27:22 GMT
postgresql_1  | 2018-06-18 10:29:14.955 GMT [35] LOG:  database system was not properly shut down; automatic recovery in progress
```

upon docker stop and docker run, postgresql does not seem to do a proper shutdown, and then has to go into recovery upon the next start.

The additional stuff I am experiencing might have to do with me mounting a cifs share, which I am not sure if the fsync issues might be caused by that fact... The database is unused at the time of stopping it.

My current docker-compose file:

```
version: "3.6"

services:
  database:
    restart: always
    image: bitnami/postgresql:latest
    environment:
      - POSTGRESQL_PASSWORD=somepassword
      - POSTGRESQL_USERNAME=gitlab
      - POSTGRESQL_DATABASE=gitlabhq_production
    networks: 
      - net_GitLab
    container_name: con_database  
    volumes:
      - Database:/bitnami

```

**Comment by juan131** [comment]
This Docker Images uses [tini](https://github.com/krallin/tini) (as you can check on the [app-entrypoint.sh](https://github.com/bitnami/bitnami-docker-postgresql/blob/master/10/rootfs/app-entrypoint.sh)) to start the services. 

Therefore, when you stop the container, *tini* is the one receiving the signal and it's the one that actually stops PostgreSQL.

@sameersbn do you know what alternative do we have to perform a "graceful stop" of PostgreSQL?

**Comment by sameersbn** [comment]
@juan131 `tini` will forward signals it receives to the `nami` process. Can you check if `nami` properly handles the signals received and properly executes the shutdown sequence for postgres? I think the shutdown sequence for postgres would have to call `pg_ctl stop` or simply forward the signal to the `postgres` process

**Comment by juan131** [comment]
@sameersbn PostgreSQL `nami` doesn't define any special method to stop PostgreSQL. Therefore, it just sends the signal to the associated process (using the PID stored in the `pidfile`)

@tompizmor do you know if have implemented custom "stop" methods on other Nami modules?

**Comment by chhetripradeep** [comment]
Hello, I think this is the reason why postgresql fails to start properly if the bitnami/postgresql container dies saying 

```
pg_ctl: another server might be running; trying to start server anyway
```
Do we know any workaround to make it start properly ? 

/cc @juan131

**Comment by chhetripradeep** [comment]
I looked into the issue more. there are two issues:

1) Currently, postmaster.pid is not removed. When bitnami/postgresql container is stopped, we need to trigger `nami stop postgresql`. In k8s world, this can be done via `preStop` hook. Once this is done, postmaster.pid is removed.

2) Currently, we are sym-linking /opt/bitnami/postgresql/data to /bitnami/postgresql/data and since a symlink has 0777 permission, postgresql fails to start throwing error:
```
data directory "/opt/bitnami/postgresql/data" has group or world access
```
I am unable to figure out where are we specifying the datadir configuration to `/opt/bitnami/postgresql/data`. Can you please help out for that.

**Comment by chhetripradeep** [comment]
Hi @juan131 - Is there a way i can change the `dataDir` from `/opt/bitnami/postgresql/data` to `/bitnami/postgresql/data` - I tried multiple stuffs like passing envvar `PGDATA` and try passing `data_directory` configuration variable, but since nami is starting postgresql specifying `-D` option - i think that is getting the highest priority among all. Can you help me in fixing this.

**Comment by chhetripradeep** [comment]
Hi @tompizmor 

Thank you for your reply.

I am using your helm chart `bitnami/postgresql` so it should be very easily reproducible. While testing some possible failure conditions, I killed the pgsql master docker container from the k8s dashboard and it didn't come up properly and only error log it printed was: ```pg_ctl: could not start server```

After spending few hours, I realized that when nami tries to bring up the pgsql server, it somehow changes the permission of datadir i.e `/opt/bitnami/postgresql/data` to `770` (i think its part of postgresql nami module) which causes failure for pgctl command throwing error ```data directory "/opt/bitnami/postgresql/data" has group or world access```

I am using 1.8.11 k8s cluster.

**Comment by chhetripradeep** [comment]
Hello @tompizmor 

Sorry I was wrong in saying that the permission change to `770` on restarting postgresql container was caused by nami. Its actually a feature of securityContext causing this 

```   
$ kubectl explain deployment.spec.template.spec.securityContext
......
FIELDS:
   fsGroup	<integer>
     A special supplemental group that applies to all containers in a pod. Some
     volume types allow the Kubelet to change the ownership of that volume to be
     owned by the pod: 1. The owning GID will be the FSGroup 2. The setgid bit
     is set (new files created in the volume will be owned by FSGroup) 3. The
     permission bits are OR'd with rw-rw---- If unset, the Kubelet will not
     modify the ownership and permissions of any volume.
......
```

I managed to workaround this permission issue by having an `initContainer` responsible to run `chmod` for the datadir. Finally everything works for me.

Thanks !!

**Comment by tompizmor** [comment]
Thanks for sharing your findings. I think it can help other users. 

As you did, an initContainer is another solution to adjust the permissions of the volume. Unfortunately, that solution is not valid for some Kubernetes distributions like Openshift that requires that all the container run with an unprivileged user. That was mainly the reason to use Kubernetes securityContexts instead of an initContainr. 
However, in this case, it seems it does not get along with PostgreSQL.

**Comment by mmack** [comment]
I had the same problem... fix with:
```
        lifecycle:
          postStart:
            exec:
              command: ["/bin/sh", "-c", "rm -f /bitnami/postgresql/data/postmaster.pid && chmod 700 /bitnami/postgresql/data"]

```

**Comment by tompizmor** [comment]
Thanks for the workaround!

**Comment by tompizmor** [comment]
We have just released a new version of the docker image that takes care of cleaning the postmaster.pid file if needed.

If you pull the latest image you should not see this issue anymore.
Please re-open this issue if you still experience the problem.

**Comment by intelfx** [comment]
@tompizmor There is also the problem with data directory permission bits, specific to k8s, that also breaks PostgreSQL when the container is restarted. Did you take a look at it? Or should I open another issue for that?

**Comment by tompizmor** [comment]
mmm, I haven't experienced the data directory permissions issue yet. Are there any specific steps to follow in order to reproduce it? 

I don't mind if you create a different issue or re-open this one

**Comment by intelfx** [comment]
@tompizmor It's k8s specific.

When pod security context is enabled/used with fsGroup set (like in the stable/postgresql Helm chart), kubelet recursively chmods the whole persistent volume as 2770 (IIRC) at each Pod startup, which upsets PostgreSQL.

**Comment by intelfx** [comment]
It's technically another problem but it has the same consequences: PostgreSQL enters a crash loop when the Pod is first restarted with persistence enabled.

**Comment by tompizmor** [comment]
I tried installing the `stable/postgresql` chart but I cannot reproduce it.

```
$ minikube version
minikube version: v0.30.0
```

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.7", GitCommit:"0c38c362511b20a098d7cd855f1314dad92c2780", GitTreeState:"clean", BuildDate:"2018-08-20T10:09:03Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:44:10Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```
```
$ helm version
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
```

Install PostgreSQL chart
```
$ helm repo update
$ helm install --name test stable/postgresql
```

Check PostgreSQL chart version
```
 $ helm list
NAME	REVISION	UPDATED                 	STATUS  	CHART           	APP VERSION	NAMESPACE
test	1       	Tue Oct 30 14:03:34 2018	DEPLOYED	postgresql-2.2.4	10.5.0     	default
```
Delete pod and check the new one starts properly. I have tried to kill the pod several times and it always worked.

**Comment by aelbarkani** [comment]
I confirm, the issue is still there. I'm deploying the last helm chart (v2.2.4), to GKE v1.9.7-gke.5 with persistence enabled.

To reproduce it just deploy the chart 2.2.4 with persistence enabled. Then delete a pod after it starts.

**Comment by intelfx** [comment]
@tompizmor Maybe it has something to do with minikube creating a cluster with some security features disabled or something? Both problems are clearly reproducible on GKE.

**Comment by tompizmor** [comment]
Yes, it is probably a minikube thing. I am going to try in GKE.

**Comment by intelfx** [comment]
I will create a new issue for clarity.

**Comment by juan131** [comment]
Cool @intelfx 

Let's continue the conversation on that issue.

**Comment by desaintmartin** [comment]
Hey guys, I would like to reopen the issue and focus on the original problem.
Each time we start a postgres, we have this "database not properly shutdown" issue. Indeed, when going into the container and doing "kill 1" (sending a sigterm to tini), in theory it should forward the signal to all children and sub-children so that postgres itself received the signal to properly shutdown.

Instead, here is what we have: 

1/ tini receives sigterm, transmit it to run.sh
2/ run.sh (and/or nami) receives sigterm, and since it has no handler, it just exits
3/ postgresql does not receive anything
4/ tini exits since its child exited
5/ tini sends sigkill to postgres ("reaping zombies") and postgresql is terminated NOT gracefully


So I have a few questions:
1/ Since we are in a container, why do we need a simple init system (tini) to manage another init system (nami) to manage postgres? Shouldn't we just exec to postgres in the end so that postgres is pid 1 (or direct child of tini if you need)?
2/ If we need to keep both tini and nami, how should we solve the problem?
 - We could use the -g option of tini which sends sigterm to all process group (basically all processes within the container)
 - we could forward the sigterm to postgresql
 - in run.sh, we could register a handler that does "nami stop postgresql" and wait until it is actually stopped

What do you think?

**Comment by tompizmor** [comment]
Hi, @desaintmartin,

Thanks a lot for the detailed information. We are already moving some of our containers to follow your first option. That is, removing 'tini' and 'nami' and simply starting the server using the run.sh command.

You can take a look at [elasticsearch](https://github.com/bitnami/bitnami-docker-elasticsearch), [mysql](https://github.com/bitnami/bitnami-docker-mysql) or [mariadb](https://github.com/bitnami/bitnami-docker-mariadb) to check this approach for example.

**Comment by desaintmartin** [comment]
Hey @tompizmor,
Do you have any ETA about this? Would you need help for this?

**Comment by tompizmor** [comment]
Hi @desaintmartin! I am afraid right now we don't have any ETA to migrate Postgresql to a full bash container but at least we can improve the startup process.

I want to take another look at the issue and try to improve the run.sh. I will update this issue in the following days.

**Comment by tompizmor** [comment]
@desaintmartin Just a quick note to let you know that we have removed `nami` from the postgresql startup process and the following warnings in the logs have disappeared:

```
postgresql_1  | 2018-06-18 10:29:14.820 GMT [35] LOG:  database system was interrupted; last known up at 2018-06-18 10:27:22 GMT
postgresql_1  | 2018-06-18 10:29:14.955 GMT [35] LOG:  database system was not properly shut down; automatic recovery in progress
```

**Comment by desaintmartin** [comment]
Great, that's a nice step forward! Thanks!

**Comment by idontusenumbers** [comment]
My postgres docker image uses a custom entry point in the form of a shell script. Here's how I solved the forwarding issue of the SIGTERM to the postgres instance:
```
# ...
echo "Starting postgres"
gosu postgres postgres  &

# Now we wait for SIGTERM sent to shell script so we can forward to postgres so it can shutdown gracefully
echo "Waiting for postmaster.pid"
while [ ! -f "$db_datapath/postmaster.pid" ]
do
  sleep 1
done
pgpid="$(head -n 1 "$db_datapath/postmaster.pid")"
killpg(){
  echo "Killing postgres pid $pgpid"
  kill "$pgpid"
}
trap killpg TERM
echo "postgres running as pid $pgpid; Waiting for SIGTERM"
while [ -f "$db_datapath/postmaster.pid" ]
do
  sleep 1
done
```

**Comment by desaintmartin** [comment]
@idontusenumbers I think this is no longer needed (at least according to my production instances correctly stopping), but it could be somehow reused for https://github.com/bitnami/bitnami-docker-postgresql/issues/200 if it makes sense.

**Comment by andresbono** [comment]
Yeah, I agree @desaintmartin. The snippet from @idontusenumbers (thanks!!) and #200 are tightly related.

**Comment by idontusenumbers** [comment]
> @idontusenumbers I think this is no longer needed
It may not be needed in the release image but as I mentioned I have additional scripting that replaces the normal script.

**Comment by andresbono** [comment]
Thanks for the tip, @idontusenumbers.
